### PR TITLE
Switch yarn to pnp linker

### PR DESCRIPTION
**Pull request by timja**
See https://yarnpkg.com/getting-started/migration#switching-to-plugnplay

More performant yarn.

Possibly in the future we could skip to zero-install which is the highest performance but let's skip that for now as it's ~30mb of zip files and I'm not sure what the appetite would be.

## Testing done

I've checked this in my local setup and in gitpod to use a from scratch environment.
I've checked views using handlebars and clicked around, also cleared browser cache.

Didn't see any issues

### Proposed changelog entries

* (not needed)
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6894"></a>

### CODE REVIEW

File: war/.yarnrc.yml
Comment (timja) [inline war/.yarnrc.yml:1]: any yarn command that modifies this file seems to remove all comments.

in this case I used https://yarnpkg.com/cli/set/version

File: war/package.json
Comment (timja) [inline war/package.json:23]: extracted from https://github.com/jenkinsci/jenkins/pull/6863/.

difficult to develop locally without as you don't know what it's complaining about, in the case of eslint it just silently exited with 1 before

File: war/package.json
Comment (timja) [inline war/package.json:70]: added by https://yarnpkg.com/cli/set/version

File: war/pom.xml
Comment (timja) [inline war/pom.xml:788]: extracted from https://github.com/jenkinsci/jenkins/pull/6863/

figured may as well given touching the same area

File: war/src/main/less/pluginSetupWizard.less
Comment (timja) [inline war/src/main/less/pluginSetupWizard.less:973]: this should not have been loading directly from node_modules, at least since webpack was used, this resolves it from the library path.

File: war/package.json
Comment (timja) [inline war/package.json:21]: stylelint custom-formatters are loaded from an absolute path which doesn't work with PnP, I did a workaround of creating a file at a known path which then loads the library correctly through module resolution:

https://github.com/stylelint/stylelint/issues/6100

**Comment by timja** [comment]
@basil requested you for review as you said you had issues with this when you tried

File: war/.yarnrc.yml
Comment (NotMyFault) [inline war/.yarnrc.yml:1]: IJ pesters me to run `yarn install` if I get a rid of node_modules but that does no longer work with pnp

File: war/package.json
Comment (NotMyFault) [inline war/package.json:15]: Same as for any following tasks:
```
/opt/homebrew/bin/yarn run lint:js
node:internal/errors:477
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_LOADER_CHAIN_INCOMPLETE]: "file:///Users/alex/gh/idea/jenkins/war/.pnp.loader.mjs 'resolve'" did not call the next hook in its chain and did not explicitly signal a short circuit. If this is intentional, include `shortCircuit: true` in the hook's return.
    at new NodeError (node:internal/errors:387:5)
    at ESMLoader.resolve (node:internal/modules/esm/loader:860:13)
    at async ESMLoader.getModuleJob (node:internal/modules/esm/loader:439:7)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:541:24)
    at async loadESM (node:internal/process/esm_loader:83:5)
    at async handleMainPromise (node:internal/modules/run_main:65:12) {
  code: 'ERR_LOADER_CHAIN_INCOMPLETE'
}

Node.js v18.6.0

Process finished with exit code 1
```
Am I missing something?

**Comment by timja** [comment]
I'll take a look at what you've seen later on Alex and see if I can reproduce.

**Comment by NotMyFault** [comment]
> I'll take a look at what you've seen later on Alex and see if I can reproduce.

Looks like my setup was scuffed, node was pointing to 18 and the package manager to npm, all good now:
Screenshot 2022-07-21 at 15 42 04

**Review by NotMyFault — APPROVED** [review]

**Comment by basil** [comment]
I made the hack even more gross in commit 33ba459242d9d91ed84f472e8fdbd4f1a295f9e4 to be able to support Renovate.

**Review by basil — APPROVED** [review]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

File: war/src/main/less/pluginSetupWizard.less
Comment (NotMyFault) [reply war/src/main/less/pluginSetupWizard.less:973]: > this should not have been loading directly from node_modules, at least since webpack was used, this resolves it from the library path.

Does it?
Screenshot 2022-07-24 at 19 22 11

Comment (timja) [reply war/src/main/less/pluginSetupWizard.less:973]: Yes with webpack you should be using the resolver to load it not hard coding a path

Comment (timja) [reply war/src/main/less/pluginSetupWizard.less:973]: bootstrap 4 switched to sass instead of less.

Comment (NotMyFault) [reply war/src/main/less/pluginSetupWizard.less:973]: > bootstrap 4 switched to sass instead of less, likely related to that.

IJ just acting up again, looks like it refuses to understand the syntax and resolves the path against the directory you are in.
In terms of the bootstrap PR, I adjusted it to `@import (css) "~bootstrap/scss/bootstrap.scss";` locally. Seems to work out well

**Pull request merged**
